nrf-hal-common/src/uarte.rs

```diff
@@ -406,18 +440,250 @@ pub enum Error {
     BufferNotInRAM,
 }
 
-pub trait Instance: Deref<Target = uarte0::RegisterBlock> + sealed::Sealed {}
+pub trait Instance: Deref<Target = uarte0::RegisterBlock> + sealed::Sealed {
+    fn ptr() -> *const uarte0::RegisterBlock;
+}
 
 mod sealed {
     pub trait Sealed {}
 }
 
 impl sealed::Sealed for UARTE0 {}
-impl Instance for UARTE0 {}
+impl Instance for UARTE0 {
+    fn ptr() -> *const uarte0::RegisterBlock {
+        UARTE0::ptr()
+    }
+}
 
 #[cfg(any(feature = "52833", feature = "52840", feature = "9160"))]
 mod _uarte1 {
     use super::*;
     impl sealed::Sealed for UARTE1 {}
-    impl Instance for UARTE1 {}
+    impl Instance for UARTE1 {
+        fn ptr() -> *const uarte0::RegisterBlock {
+            UARTE1::ptr()
+        }
+    }
+}
+
+/// Interface for the TX part of a UART instance that can be used independently of the RX part.
+pub struct UarteTx<T>
+where
+    T: Instance,
+{
+    _marker: core::marker::PhantomData<T>,
+    tx_buf: &'static mut [u8],
+    written: usize,
+}
+
+/// Interface for the RX part of a UART instance that can be used independently of the TX part.
+pub struct UarteRx<T>
+where
+    T: Instance,
+{
+    _marker: core::marker::PhantomData<T>,
+    rx_buf: &'static mut [u8],
+}
+
+impl<T> UarteTx<T>
+where
+    T: Instance,
+{
+    fn new(tx_buf: &'static mut [u8]) -> Result<UarteTx<T>, Error> {
+        if tx_buf.len() == 0 {
+            return Err(Error::TxBufferTooSmall);
+        }
+
+        if tx_buf.len() > EASY_DMA_SIZE {
+            return Err(Error::TxBufferTooLong);
+        }
+
+        Ok(UarteTx {
+            _marker: core::marker::PhantomData,
+            tx_buf,
+            written: 0,
+        })
+    }
+}
+
+impl<T> UarteRx<T>
+where
+    T: Instance,
+{
+    fn new(rx_buf: &'static mut [u8]) -> Result<UarteRx<T>, Error> {
+        if rx_buf.len() == 0 {
+            return Err(Error::RxBufferTooSmall);
+        }
+
+        if rx_buf.len() > EASY_DMA_SIZE {
+            return Err(Error::RxBufferTooLong);
+        }
+
+        Ok(UarteRx {
+            _marker: core::marker::PhantomData,
+            rx_buf,
+        })
+    }
+}
+
+impl<T> Drop for UarteTx<T>
+where
+    T: Instance,
+{
+    fn drop(&mut self) {
+        let uarte = unsafe { &*T::ptr() };
+
+        let in_progress = uarte.events_txstarted.read().bits() == 1;
+        // Stop any ongoing transmission
+        if in_progress {
```

Comment:
Shouldn't we reset `events_txstarted` in this `if` block? Otherwise it seems to me that it'd stay on indefinitely